bitcoin/src/bip32.rs

```diff
@@ -730,33 +1076,73 @@ impl Xpub {
     /// the internal public key representation.
     pub fn to_x_only_pub(self) -> XOnlyPublicKey { XOnlyPublicKey::from(self.public_key) }
 
+    /// Derive an extended public key from a path that contains NormalChildNumbers.
+    pub fn derive_pub<
+        'a,
+        C: secp256k1::Verification,
+        T: IntoIterator<Item = &'a NormalChildNumber>,
+    >(
+        &self,
+        secp: &Secp256k1<C>,
+        path: T,
+    ) -> Xpub {
+        let mut pk: Xpub = *self;
+        for normal_cnum in path.into_iter() {
+            pk = pk.internal_derive_pub(secp, normal_cnum.to_number()).unwrap();
+        }
+        pk
+    }
+
     /// Attempts to derive an extended public key from a path.
     ///
-    /// The `path` argument can be any type implementing `AsRef<ChildNumber>`, such as `DerivationPath`, for instance.
-    pub fn derive_pub<C: secp256k1::Verification, P: AsRef<[ChildNumber]>>(
+    /// The `path` argument can be any type implementing `IntoIterator<Item = &'a ChildNumber>`, such as `DerivationPath`, for instance.
+    pub fn try_derive_pub<
+        'a,
+        C: secp256k1::Verification,
+        P: IntoIterator<Item = &'a ChildNumber>,
+    >(
         &self,
         secp: &Secp256k1<C>,
-        path: &P,
+        path: P,
     ) -> Result<Xpub, Error> {
         let mut pk: Xpub = *self;
-        for cnum in path.as_ref() {
-            pk = pk.ckd_pub(secp, *cnum)?
+        for normal_cnum in path.into_iter() {
+            pk = pk.internal_derive_pub(secp, *normal_cnum)?
         }
         Ok(pk)
     }
 
+    /// Public->Public child key derivation
+    fn internal_derive_pub<C: secp256k1::Verification, N: TryInto<ChildNumber>>(
+        &self,
+        secp: &Secp256k1<C>,
+        number_from: N,
+    ) -> Result<Xpub, Error> {
+        let number_from = number_from.try_into()?;
+        let (sk, chain_code) = self.ckd_pub_tweak(number_from)?;
+
+        Ok(Xpub {
+            network: self.network,
+            depth: self.depth + 1,
+            parent_fingerprint: self.fingerprint(),
+            child_number: number_from,
+            public_key: tweaked,
+            chain_code,
+        })
+    }
+
```

Comment:
All the Cis are failing due `try_into()`. Because it doesnt have a specific return error.

Reply:
`N: Into<ChildNumber>` was working fine.